### ISSUE TITLE
[REGRESSION]: Azure file copy fails on one Agent in Pool

### ISSUE THREAD

**Issue by lucavoit** [microsoft/azure-pipelines-tasks#20041]
### New issue checklist

- [X] I searched for [existing GitHub issues](https://github.com/microsoft/azure-pipelines-tasks/issues)
- [X] I read [pipeline troubleshooting guide](https://docs.microsoft.com/vsts/build-release/actions/troubleshooting)
- [X] I checked [how to collect logs](https://learn.microsoft.com/azure/devops/pipelines/troubleshooting/review-logs?view=azure-devops)

### Task name

Azure file copy

### Breaking task version

6.*

### Last working task version

5.*

### Regression Description

We recently updated our azure service connection to use Workload identity federation, which broke our pipeline using v5.* so we changed to v6. This fixed the issues in our pipelines except for one pipeline which runs on a specific self hosted agent.
On this agent the upload task fails with: `Failed to perform Auto-login: PSContextCredential[31;1mGet-AzAccessToken: -REDACTED-;1mObject reference not set to an instance of an object.[0m` 
If I run the same task in a mini repro pipeline on any different agent in our pool, the task works fine.

In trying to fix this issue we updated the agent and all the powershell modules to the latest version - no difference.
Any help would be greatly appreciated. Thanks

### Environment type (Please select at least one enviroment where you face this issue)

- [X] Self-Hosted
- [ ] Microsoft Hosted
- [ ] VMSS Pool
- [ ] Container

### Azure DevOps Server type

dev.azure.com (formerly visualstudio.com)

### Operation system

Windows 11

### Relevant log output

```shell
2024-06-19T11:15:14.1507191Z INFO: Scanning...
2024-06-19T11:15:15.5513209Z 
2024-06-19T11:15:15.5514219Z Failed to perform Auto-login: PSContextCredential[31;1mGet-AzAccessToken: -REDACTED-;1mObject reference not set to an instance of an object.[0m
2024-06-19T11:15:15.5514447Z .
2024-06-19T11:15:15.5639767Z ##[debug]ExceptionMessage: AzCopy.exe exited with non-zero exit code while uploading files to blob storage.
2024-06-19T11:15:15.5816208Z ##[debug]Processed: ##vso[task.logissue type=error;code={"Task_Internal_Error":"BlobUploadFailed"};]
```

**Comment by v-schhabra** [comment]
Hi @lucavoit 
Thanks for reporting the issue. Could you please share the complete pipeline logs by adding variable system.debug to "true"?

**Comment by v-bsanthanak** [comment]
Duplicate of #20003

**Comment by lucavoit** [comment]
> Duplicate of #20003

@v-bsanthanak I'm sorry, but closing this issue as a duplicate of another issue (which has an entirely different error message) is hardly a satisfactory solution? Is there any info on how to fix this on our side or if you will fix this with an update?

**Comment by lucavoit** [comment]
@v-bsanthanak @v-schhabra Anymore info on this issue? We really need a solution / fix for this

**Comment by v-schhabra** [comment]
> @v-bsanthanak @v-schhabra Anymore info on this issue? We really need a solution / fix for this

We have created a PR for this issue. Once it is merged and deployed will update here.

**Comment by lucavoit** [comment]
@v-schhabra could you maybe link the relevant PR or give some more updates?

**Comment by v-schhabra** [comment]
Hi @lucavoit 
Thanks for following up. 
PR has been created but it is waiting for the deployments to complete.
https://github.com/microsoft/azure-pipelines-tasks/pull/20117

**Comment by lucavoit** [comment]
@v-schhabra this is still not working for us, even when using the new task version 6.242.10. Is there anything else we need to do / update?

**Comment by v-schhabra** [comment]
@lucavoit 
Could you pls share the complete logs by enabling system.debug to true?

**Comment by v-schhabra** [comment]
If the logs are same as earlier then it is fine. I just wanted to compare the logs with another cx who is also having same issue.

**Comment by lucavoit** [comment]
@v-schhabra yeah I see no relevant difference. Do you need anymore information from our side to investigate this further?

**Comment by v-schhabra** [comment]
> @v-schhabra yeah I see no relevant difference. Do you need anymore information from our side to investigate this further?

As of now we dont need any info. Will post here if I need anything.

**Comment by lucavoit** [comment]
An interesting find: the upload task only fails in one specific release pipeline - there is a build pipeline also running on the same machine / agent that includes a blob upload which does work

**Comment by v-schhabra** [comment]
We observed that if we are using VM in that specific case pipeline is erroring out otherwise (blob) it is working fine.

**Comment by lucavoit** [comment]
in our case the agent runs on a physical machine

**Comment by v-schhabra** [comment]
@lucavoit can you try to create a new release and check if the issue persist?
We observed that the issue is not happening from task side and from the logs we could see the error is throwing from Azcopy.exe.
So, can you pls create a new release pipeline and check if this task works fine?

**Comment by lucavoit** [comment]
> @lucavoit We observed that the issue is not happening from task side and from the logs we could see the error is throwing from Azcopy.exe. So, can you pls create a new release pipeline and check if this task works fine?

does it need to be a new pipeline for it to work again?

**Comment by lucavoit** [comment]
@v-schhabra I have a small update: we manually upgraded the azcopy version from 10.25.1 to 10.26.0 since we found a log message regarding this in the latest run. This however did not change the outcome. However we are yet to test a new pipeline as you suggested, as we are a bit apprehensive about the effort it would take us considering our current workload

**Comment by v-schhabra** [comment]
@lucavoit please share the logs for further investigation?

**Comment by lucavoit** [comment]
@v-schhabra yes, we will provide the logs later today

**Comment by lucavoit** [comment]
@v-schhabra any updates?

**Comment by lucavoit** [comment]
Well @v-schhabra I did some more digging and I have an update for you:

I made a minimal pipeline setup for debug purposes, using a Azure CLI task to execute my own powershell script.
Using the following script I got it to work:

`
$env:AZCOPY_AUTO_LOGIN_TYPE="PSCRED"
$env:AZCOPY_TENANT_ID="<tenant-id>"
Get-AzContext
azcopy.exe copy ......
`

with the Get-AzContext being vital for it to work properly, before that I ran into the exact same error message I got as using your task - so maybe this could be a solution here to?

**Comment by v-schhabra** [comment]
Hi @lucavoit 
Thanks for the update.
